### ISSUE TITLE
Streams evaluation order (feature or bug?)

### ISSUE THREAD

**Issue by fadado** [stedolan/jq#1283]
The evaluation order for streams (generators) in **jq** is different for cases like these

    $ jq -cn '[(1,3) + (2,3,4,5)]'
    [3,5,4,6,5,7,6,8]
    $ jq -cn 'def sum(a;b): a + b; [sum((1,3); (2,3,4,5))]'
    [3,5,4,6,5,7,6,8]

and for pipes

    $ jq -cn '[(1,3) as $x | (2,3,4,5) as $y | $x + $y]'
    [3,4,5,6,5,6,7,8]

In the [Icon language](https://www.cs.arizona.edu/icon/docs/docs.htm), one of the direct inspirations for **jq**, the order is always the same (left to rigth). This paragraph from one [Icon paper](https://www.cs.arizona.edu/icon/ftp/doc/tr82_8a.pdf)  describes the Icon behavior:

>In the absence of control structures, the arguments of a function call are
>evaluated from left to right. If evaluation of any argument expression fails,
>the last argument expression to be evaluated is resumed to produce another
>result. If it is depleted and does not produce a result, the next previous
>argument expression is resumed, and so on. When a resumed argument expression
>produces a result, the remaining argument expressions to the right are
>evaluated again. If the first argument expression fails to produce a result,
>the function is not invoked and the entire expression fails. If all the
>arguments produce a result, the function is invoked with those argument values.
>If the function fails for these argument values, the last argument expression
>is resumed for another result and evaluation of the argument expressions
>proceeds as described above. Thus generators are resumed in a last-in,
>first-out manner.

My question is, the dual **jq** behavior is a design feature or not?
I like very much the development of Icon ideas inside a pure functional framework, and I think the inspiration must be deepened, for example including the missing _dot product_ (`zip`) to **jq**.

**Comment by slapresta** [comment]
For what it's worth, `zip` is just called `transpose` in jq:

```
jq -nc '[[1, 2, 3], [4, 5, 6]] | transpose'
[[1,4],[2,5],[3,6]]
```

**Comment by fadado** [comment]
Yes, I know, but **jq** `transpose`  expects an array of arrays as input, not streams. **jq**, like Icon, lacks dot product on streams; only cross product is available using pipes. The Icon solution is weird and uses co-expressions; for **jq** I will like a kind of parallel  `reduce` (_fold_) and/or `foreach` (_scan_), but a family of _zip_ and _zipWith_ functions operating on streams  like in Haskell will be also ok.

I can define a zip on streams like

```
def zip(g; h):
    [[g], [h]] as $pair
    | ($pair | map(length) | max) as $longest
    | range($longest)
    | [$pair[0][.], $pair[1][.]]
;
```

but only works on finite streams and has to build the whole arrays in memory!
As an illustration, the following example has to convert streams to arrays for
parallel display of equivalent codes:

```
#!/usr/bin/jq -nRrf

def chars: (./"")[];

def octcode:
    ("01"|chars) as $d2 |
    ("01234567"|chars) as $d1 |
    ("01234567"|chars) as $d0 |
    $d2+$d1+$d0
;

def deccode:
    label $pipe |
    ("01"|chars) as $d2 |
    ("0123456789"|chars) as $d1 |
    ("0123456789"|chars) as $d0 |
    if $d2 == "1" and $d1 == "2" and $d0 == "8"
    then break $pipe
    else $d2+$d1+$d0 end
;

def hexcode:
    ("01234567"|chars) as $d1 |
    ("0123456789ABCDEF"|chars) as $d0 |
    $d1+$d0
;

[ [octcode], [deccode], [hexcode] ] | transpose
```

Icon can process in parallel the three streams, as shown in
https://www.cs.arizona.edu/icon/ftp/doc/tr82_8a.pdf (page 17).

**Comment by pkoppstein** [comment]
> Is the dual jq behavior a design feature or not?

It is as intended.  It has to do with the backtracking points, and the fact that "$-variables" hold JSON entities, not closures.

As for `zip/2`, the following definition will work on finite and infinite streams, though with dubious efficiency:

    def zip(f; g):
      def butfirst(s): foreach s as $i (0; .+1; if . > 1 then $i else empty end);
      first(f) as $f
      | first(g) as $g
      | [$f, $g], zip( butfirst(f);  butfirst(g));

You can test it using:

     def integers: .+1 | (., integers);
     zip(integers; integers)

**Comment by fadado** [comment]
> Is the dual jq behavior a design feature or not?
>> It is as intended. It has to do with the backtracking points, and the fact that "$-variables" hold JSON entities, not closures.
 
Thank you for the fast response. I actually like the availability of two evaluation strategies to choose, one with the rigth gear running faster (`g as $a | h as $b | f($a; $b)`) and one with a faster left (`f(g; h)`).

> As for zip/2, the following definition will work on finite and infinite streams, though with dubious efficiency:
...

Yes. I use this _full stream copy_ for implementing `drop` or `rest`, until better solutions exists:

```
def drop($n; g):
    select($n >= 0) |
    if $n == 0
    then g
    else
        foreach g as $item ($n; .-1;  # . never will arrive to -infinite...
            if . < 0 then $item else empty end)
    end
; 
def rest(g):
    drop(1; g)
;
```

By the way, in this same stream processing area, the `isempty`implementation is fast:

```
def isempty(g):
    (label $block | g | 1, break $block)//0
    | .==0  # count is 0
;
```

**Comment by pkoppstein** [comment]
@fadado - I've added `def isempty(s)` along the lines of your definition to the jq FAQ. Thanks!

**Comment by fadado** [comment]
> @fadado - I've added def isempty(s) along the lines of your definition to the jq FAQ. Thanks!

Is an honor, thank you. I will close this issue now, and start a new one for a behavior of `range` I can't understand.

JJOR

**Comment by nicowilliams** [comment]
@fadado I think what's happening is that "left" and "right" are confused in the case of binary operators.  I'm actually not entirely sure where that's happening.  I set a breakpoint on `gen_binop()` when parsing that program and I see that the left argument is the correct one, but when I `--debug-dump-disasm` I see that the two are reversed...  Ah, maybe it's from this:

```
      case CLOSURE_CREATE_C: {
        for (inst* i; (i = block_take(&curr->arglist)); ) {
          assert(i->op == CLOSURE_CREATE); // FIXME
          block body = i->subfn;
          i->subfn = gen_noop();
          inst_free(i);
          // arguments should be pushed in reverse order, prepend them to prelude
          errors += expand_call_arglist(&body, args, env);
          prelude = BLOCK(gen_subexp(body), prelude);
          actual_args++;
        }
```

"arguments should be pushed in reverse order, prepend them to prelude" -- sounds like that might be it.

Compare to:

```
$ jq -cn 'def f($a;$b):$a+$b; [f(10,30;2,3,4,5)]'
[12,13,14,15,32,33,34,35]
$ 
```

I'll have to ask @stedolan if this is really correct!  To me it seems rather wrong.  Mind you, it's not like we can change it now, I don't think.  It probably wouldn't surprise too many users, but it might nonetheless surprise some and break their programs, so I'm inclined to just document this.
